sympy/functions/special/gamma_functions.py

```diff
@@ -1009,3 +1009,64 @@ def trigamma(x):
     .. [3] http://functions.wolfram.com/GammaBetaErf/PolyGamma2/
     """
     return polygamma(1, x)
+
+def multivariate_gamma(x, p):
+    r"""
+    The multivariate gamma function is a generalization of the gamma function i.e,
+
+    .. math::
+        \Gamma_p(z) = \pi^{p(p-1)/4}\prod_{k=1}^p \Gamma[z + (1 - k)/2].
+
+    Special case, multivariate_gamma(x, 1) = gamma(x)
+
+    Parameters
+    ==========
+
+    p: order or dimension of the multivariate gamma function
+
+    Examples
+    ========
+
+    >>> from sympy import S, I, pi, oo, gamma, multivariate_gamma
+    >>> from sympy import Symbol
+    >>> x = Symbol('x')
+    >>> p = Symbol('p', positive=True, integer=True)
+
+    >>> multivariate_gamma(x, p)
+    pi**(p*(p - 1)/4)*Product(gamma(-k/2 + x + 1/2), (k, 1, p))
+
+    Several special values are known:
+    >>> multivariate_gamma(1, 1).doit()
+    1
+    >>> multivariate_gamma(4, 1).doit()
+    6
+    >>> multivariate_gamma(S(3)/2, 1).doit()
+    sqrt(pi)/2
+
+    Writing multivariate_gamma in terms of gamma function
+    >>> multivariate_gamma(x, 1).doit()
+    gamma(x)
+
+    >>> multivariate_gamma(x, 2).doit()
+    sqrt(pi)*gamma(x)*gamma(x - 1/2)
+
+    >>> multivariate_gamma(x, 3).doit()
+    pi**(3/2)*gamma(x)*gamma(x - 1)*gamma(x - 1/2)
+
+    See Also
+    ========
+
+    gamma, lowergamma, uppergamma, polygamma, loggamma, digamma, trigamma
+    sympy.functions.special.beta_functions.beta
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Multivariate_gamma_function
+    """
+    psym = sympify(p)
+    if fuzzy_not(fuzzy_and([psym.is_integer, psym.is_positive])):
+        raise TypeError('p must be a positive integer')
+    from sympy import Product
+    from sympy.abc import k
+    return pi**(psym*(psym-1)/4)*Product(gamma(x + (1 - k)/2), (k, 1, psym))
```

Comment:
Lower case methods return evaluated results when possible, that is they shouldn't require `doit`.
That said, unevaluated expressions can still be obtained by passing `evaluate=False`.
```py
>>> gamma(5)
24
>>> gamma(S(5)/2)
3*sqrt(pi)/4
>>> gamma(7, evaluate=False)
gamma(7)
```

Reply:
What do think about the following diff:
```
--- a/sympy/functions/special/gamma_functions.py
+++ b/sympy/functions/special/gamma_functions.py
@@ -1010,7 +1010,7 @@ def trigamma(x):
     """
     return polygamma(1, x)
 
-def multivariate_gamma(x, p):
+def multivariate_gamma(x, p, evaluate=True):
     r"""
     The multivariate gamma function is a generalization of the gamma function i.e,
 
@@ -1065,8 +1065,18 @@ def multivariate_gamma(x, p):
     .. [1] https://en.wikipedia.org/wiki/Multivariate_gamma_function
     """
     psym = sympify(p)
-    if fuzzy_not(fuzzy_and([psym.is_integer, psym.is_positive])):
+    xsym = sympify(x)
+    if psym.is_integer is False or psym.is_positive is False:
         raise TypeError('p must be a positive integer')
     from sympy import Product
     from sympy.abc import k
+    if evaluate:
+        if psym.is_Integer and xsym.is_Integer:
+            if 2*x + 1 > p:
+                return (pi**(psym*(psym-1)/4)*Product(gamma(x + (1 - k)/2),
+                (k, 1, psym))).doit()
+            return pi**(psym*(psym-1)/4)*Product(gamma(x + (1 - k)/2),
+            (k, 1, psym))
+        return (pi**(psym*(psym-1)/4)*Product(gamma(x + (1 - k)/2),
+        (k, 1, psym))).doit()
     return pi**(psym*(psym-1)/4)*Product(gamma(x + (1 - k)/2), (k, 1, psym))
```
Example usage:
```
>>> from sympy import *
>>> mg = multivariate_gamma
>>> mg(1, 1)
1
>>> mg(1, 1, evaluate=False)
Product(gamma(3/2 - k/2), (k, 1, 1))
>>> x, p = symbols('x p')
>>> mg(x, p)
pi**(p*(p - 1)/4)*Product(gamma(-k/2 + x + 1/2), (k, 1, p))
>>> mg(x, 1)
gamma(x)
>>> mg(1, p)
pi**(p*(p - 1)/4)*Product(gamma(3/2 - k/2), (k, 1, p))
```

Reply:
I think `multivariate_gamma` could be implemented along the lines of `gamma` inheriting from `Function` (and not `digamma` or `trigamma` which are just Python functions based off `polygamma`).

Reply:
@sidhantnagpal I am trying to implement multivariate gamma function along the lines of `gamma` class. I found the following bug with gamma function.
```
>>> gamma(-1) # -1 is pole for gamma function
zoo
>>> gamma(-1.0)
Traceback (most recent call last):
  File "/home/ritesh/Desktop/sympy/sympy/core/evalf.py", line 1308, in evalf
    rf = evalf_table[x.func]
KeyError: gamma

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ritesh/Desktop/sympy/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/ritesh/Desktop/sympy/sympy/core/function.py", line 463, in __new__
    result = result.evalf(mlib.libmpf.prec_to_dps(pr))
  File "/home/ritesh/Desktop/sympy/sympy/core/evalf.py", line 1442, in evalf
    result = evalf(self, prec + 4, options)
  File "/home/ritesh/Desktop/sympy/sympy/core/evalf.py", line 1314, in evalf
    xe = x._eval_evalf(prec)
  File "/home/ritesh/Desktop/sympy/sympy/core/function.py", line 586, in _eval_evalf
    v = func(*args)
  File "/home/ritesh/venv/lib/python3.6/site-packages/mpmath/ctx_mp_python.py", line 1012, in f
    return ctx.make_mpf(mpf_f(x._mpf_, prec, rounding))
  File "/home/ritesh/venv/lib/python3.6/site-packages/mpmath/libmp/gammazeta.py", line 1954, in mpf_gamma
    raise ValueError("gamma function pole")
ValueError: gamma function pole
```
I have opened this [issue](https://github.com/sympy/sympy/issues/16847) for this. After this is sorted, we can `multivariate_gamma` will work even for negative numbers. Meanwhile, I will share my code here, for you to review.